Client/Frontend/Widgets/AutocompleteTextField.swift

```diff
@@ -0,0 +1,139 @@
+    func autocompleteTextFieldDidBeginEditing(autocompleteTextField: AutocompleteTextField)
+    func autocompleteTextFieldDidEndEditing(autocompleteTextField: AutocompleteTextField)
+}
+
+/// Data source for the text field autocompletion.
+protocol AutocompleteTextFieldDataSource: class {
+    func autocompleteTextField(autocompleteTextField: AutocompleteTextField, suggestionForText text: String, complete: String? -> Void)
+}
+
+private struct AutocompleteTextFieldUX {
+    static let HighlightColor = UIColor(rgb: 0xccdded)
+}
+
+class AutocompleteTextField: UITextField, UITextFieldDelegate {
+    weak var autocompleteDelegate: AutocompleteTextFieldDelegate?
+    weak var autocompleteDataSource: AutocompleteTextFieldDataSource?
+
+    private var autocompleting = false
+
+    override init() {
+        super.init()
+    }
+
+    override init(frame: CGRect) {
+        super.init(frame: frame)
+        super.delegate = self
+    }
+
+    required init(coder aDecoder: NSCoder) {
+        super.init(coder: aDecoder)
+        super.delegate = self
+    }
+
+    private var enteredText: NSString = "" {
+        didSet {
+            autocompleteDelegate?.autocompleteTextField(self, didTextChange: enteredText)
+        }
+    }
+
+    func textField(textField: UITextField, shouldChangeCharactersInRange range: NSRange, replacementString string: String) -> Bool {
+        let text = textField.text as NSString
+
+        // If we're autocompleting, stretch the range to the end of the text.
+        let range = autocompleting ? NSMakeRange(range.location, text.length - range.location) : range
+
+        // Update the entered text if we're adding characters or we're not autocompleting.
+        // Otherwise, we're going to clear the autocompletion, which means the entered text shouldn't change.
+        if !string.isEmpty || !autocompleting {
+            enteredText = text.stringByReplacingCharactersInRange(range, withString: string)
+        }
+
+        // TODO: short-circuit if no previous match and text is longer.
+        // TODO: cache last autocomplete result to prevent unnecessary iterations.
+        if !string.isEmpty {
+            autocompleteDataSource?.autocompleteTextField(self, suggestionForText: enteredText) { suggestion in
+                // If there's no suggestion, clear the existing autocompletion and bail.
+                if suggestion == nil {
+                    self.clearAutocomplete()
+                    return
+                }
+
+                // Create the attributed string with the autocompletion highlight.
+                let attributedString = NSMutableAttributedString(string: suggestion!)
+                attributedString.addAttribute(NSBackgroundColorAttributeName, value: AutocompleteTextFieldUX.HighlightColor, range: NSMakeRange(self.enteredText.length, countElements(suggestion!) - self.enteredText.length))
+                textField.attributedText = attributedString
```

Comment:
I guess if you're using didSet, you'll hit loops here where you'll have to be careful to bail earlier.